ecommerce/api_test.py

```diff
@@ -407,3 +350,143 @@ def test_get_product_price(basket_and_coupons):
         .price
     )
     assert expected_price == get_product_price(basket_and_coupons.basket_item.product)
+
+
+@pytest.mark.parametrize("has_coupon", [True, False])
+def test_get_product_version_price_with_discount(has_coupon, basket_and_coupons):
+    """
+    get_product_version_price_with_discount should check if the coupon exists and if so calculate price based on its
+    discount.
+    """
+    product_version = basket_and_coupons.basket_item.product.productversions.order_by(
+        "-created_on"
+    ).first()
+    product_price = product_version.price
+
+    coupon_version = basket_and_coupons.coupongroup_best.coupon_version
+    discount = coupon_version.invoice_version.amount
+    price = get_product_version_price_with_discount(
+        coupon_version=coupon_version if has_coupon else None,
+        product_version=product_version,
+    )
+    if has_coupon:
+        assert price == product_price * (1 - discount)
+    else:
+        assert price == product_price
+
+
+def test_get_new_order_by_reference_number(basket_and_coupons):
+    """
+    get_new_order_by_reference_number returns an Order with status created
+    """
+    user = basket_and_coupons.basket_item.basket.user
+    order = create_unfulfilled_order(user)
+    same_order = get_new_order_by_reference_number(make_reference_id(order))
+    assert same_order.id == order.id
+
+
+def test_parse():
+    """
+    Test parse errors are handled well
+    """
+    with pytest.raises(ParseException) as ex:
+        get_new_order_by_reference_number("XYZ-1-3")
+    assert ex.value.args[0] == "Reference number must start with MITXPRO-"
+
+    with pytest.raises(ParseException) as ex:
+        get_new_order_by_reference_number("MITXPRO-no_dashes_here")
+    assert ex.value.args[0] == "Unable to find order number in reference number"
+
+    with pytest.raises(ParseException) as ex:
+        get_new_order_by_reference_number("MITXPRO-something-NaN")
+    assert ex.value.args[0] == "Unable to parse order number"
+
+    with pytest.raises(ParseException) as ex:
+        get_new_order_by_reference_number("MITXPRO-not_matching-3")
+    assert ex.value.args[0] == "CyberSource prefix doesn't match"
+
+
+def test_status(basket_and_coupons):
+    """
+    get_order_by_reference_number should only get orders with status=CREATED
```

Comment:
Sorry, the function names here came copying and pasting from micromasters, where the functions were part of a class that was more descriptive. I'll update the names and `get_order_by_reference_number`

This function shouldn't assert the status but instead check that the proper exception is raised when the order is missing. I'll change the test name and also make sure that the status is checked in some test.

Reply:
So it looks like the docstring is wrong here. In https://github.com/mitodl/micromasters/commit/bf9dc6c7c3f70aa808af8c712b3d3485f12f73de I removed the filter on `status` because we wanted to handle duplicate order fulfillments gracefully.